### DISCUSSION TITLE
Add data channel to WebRTC ?

### DISCUSSION THREAD

**Discussion by Kiwa21** [bluenviron/mediamtx#1456] [Questions]
Hello @aler9  !
Thanks for adding the WebRTC feature, it works like a charm.

Do you know if it's possible to pipe a data track to the WebRTC data channel ?
For example to forward MPEG-TS with video and data to WebRTC and consume data with javascript on the webclient side ?
Thanks !

**Comment by aler9** [comment]
Hello Gregoire, of course it's theoretically possible, the server has to be improved in order to recognize data tracks sent with RTP/MPEG-TS, UDP/MPEG-TS or RTP. Once the server is able to recognize data tracks, it can forward them to different protocols.

Much of the work consists into finding specifications that define how a data track is structured in both MPEG-TS and WebRTC.

Once the specifications are clear, practical work is (in my opinion) easy:

1) define a format dedicated to data tracks

2) improve the function that detects MPEG-TS tracks, in order to detect data tracks too and return the format associated with data tracks:

https://github.com/aler9/rtsp-simple-server/blob/633340bcbb54c366472eeaa905287a51da172409/internal/hls/mpegts/tracks.go#L46

3) edit the WebRTC handler in order to link data tracks to output clients:

https://github.com/aler9/rtsp-simple-server/blob/633340bcbb54c366472eeaa905287a51da172409/internal/core/webrtc_conn.go#L541

and that should be all.

[a comment by an automated account is not shown]
